pyomo/contrib/satsolver/satsolver.py

```diff
@@ -0,0 +1,273 @@
+
+    # processes pyomo model into SMT model
+    def _process_model(self, model):
+        for v in model.component_data_objects(ctype=Var, descend_into=False):
+            smtstring = self.add_var(v)
+        for c in model.component_data_objects(ctype=Constraint, active=True):
+            self.add_expr(c.expr)
+        for djn in model.component_data_objects(ctype=Disjunction):
+            if djn.active:
+                self._process_active_disjunction(djn)
+            else:
+                self._process_inactive_disjunction(djn)
+
+    # define bound constraints
+    def _add_bound(self, var):
+        nm = self.variable_label_map.getSymbol(var)
+        lb = var.lb
+        ub = var.ub
+        if lb is not None:
+            self.bounds_list.append("(assert (>= " + nm + " " + str(lb) + "))\n")
+        if ub is not None:
+            self.bounds_list.append("(assert (<= " + nm + " " + str(ub) + "))\n")
+
+    # define variables
+    def add_var(self, var):
+        label = self.variable_label_map.getSymbol(var)
+        domain = type(var.domain)
+        if domain is RealSet:
+            self.variable_list.append("(declare-fun " + label + "() Real)\n")
+            self._add_bound(var)
+        elif domain is IntegerSet:
+            self.variable_list.append("(declare-fun " + label + "() Int)\n")
+            self._add_bound(var)
+        elif domain is BooleanSet:
+            self.variable_list.append("(declare-fun " + label + "() Int)\n")
+            self._add_bound(var)
+        else:
+            raise NotImplementedError("SMT cannot handle" + str(domain) + "variables")
+        return label
+
+    # Defines SMT expression from pyomo expression
+    def add_expr(self, expression):
+        smtexpr = self.walker.walk_expression(expression)
+        self.expression_list.append("(assert " + smtexpr + ")\n")
+
+    # Computes the SMT Model for the disjunction from the internal class storage
+    def _compute_disjunction_string(self, smt_djn):
+        djn_string = smt_djn[0]
+        for disj in smt_djn[1]:
+            cons_string = "true"
+            for c in disj[1]:
+                cons_string = "(and " + cons_string + " " + c + ")"
+            djn_string = djn_string + "(assert (=> ( = 1 " + disj[0] + ") " + cons_string + "))\n"
+        return djn_string
+
+    # converts disjunction to internal class storage
+    def _process_active_disjunction(self, djn):
+        xor_expr = "0"
+        disjuncts = []
+        for disj in djn.disjuncts:
+            constraints = []
+            iv = disj.indicator_var
+            label = self.add_var(iv)
+            xor_expr = "(+ " + xor_expr + " " + label + ")"
+            for c in disj.component_data_objects(ctype=Constraint, active=True):
+                constraints.append(self.walker.walk_expression(c.expr))
+            disjuncts.append((label, constraints))
+        xor_expr = "(assert (= 1 " + xor_expr + "))\n"
+        self.disjunctions_list.append((xor_expr, disjuncts))
```

Comment:
Note that the disjunction is only an XOR if the `djn.xor` is True.  Otherwise it is a simple OR.